crates/autopilot/src/domain/events/mod.rs

```diff
@@ -0,0 +1,205 @@
+        let mut settlements = Vec::new();
+        let mut trades = Vec::new();
+        let mut cancellations = Vec::new();
+        let mut presignatures = Vec::new();
+
+        for event in events {
+            let metadata = match event.meta {
+                Some(meta) => meta,
+                None => {
+                    tracing::warn!(?event, "GPv2Contract event without meta data");
+                    continue;
+                }
+            };
+            match event.data {
+                boundary::events::GPv2Contract::Settlement(event) => {
+                    settlements.push(settlement::Settlement {
+                        block_number: metadata.block_number,
+                        log_index: metadata.log_index,
+                        tx_hash: metadata.transaction_hash,
+                        solver: event.solver,
+                    });
+                }
+                boundary::events::GPv2Contract::Trade(event) => {
+                    trades.push(settlement::Trade {
+                        block_number: metadata.block_number,
+                        log_index: metadata.log_index,
+                        uid: uid(&event.order_uid.0)?,
+                        sell_amount_including_fee: event.sell_amount,
+                        buy_amount: event.buy_amount,
+                        fee_amount: event.fee_amount,
+                    });
+                }
+                boundary::events::GPv2Contract::OrderInvalidated(event) => {
+                    cancellations.push(settlement::Cancellation {
+                        block_number: metadata.block_number,
+                        log_index: metadata.log_index,
+                        uid: uid(&event.order_uid.0)?,
+                    });
+                }
+                boundary::events::GPv2Contract::PreSignature(event) => {
+                    presignatures.push(settlement::PreSignature {
+                        block_number: metadata.block_number,
+                        log_index: metadata.log_index,
+                        uid: uid(&event.order_uid.0)?,
+                        owner: event.owner,
+                        signed: event.signed,
+                    });
+                }
+                boundary::events::GPv2Contract::Interaction(_) => {}
+            }
+        }
+
+        self.persistence
+            .store_settlement_events(tx, settlements)
+            .await
+            .map_err(Error::Persistence)?;
+        self.persistence
+            .store_trade_events(tx, trades)
+            .await
+            .map_err(Error::Persistence)?;
+        self.persistence
+            .store_cancellation_events(tx, cancellations)
+            .await
+            .map_err(Error::Persistence)?;
+        self.persistence
+            .store_presignature_events(tx, presignatures)
+            .await
+            .map_err(Error::Persistence)
+    }
+}
+
+pub fn uid(bytes: &[u8]) -> Result<domain::OrderUid, Error> {
```

Comment:
This is definitely possible, but would require defining another domain error type on OrderUid which then needs to get converted into the events domain error type. Quite a lot of boiler plate for 0 code reuse so far honestly (but sure can move).

Reply:
Yeah. I think I got the responsibilities confused. The `domain` should probably not concern itself with turning possibly incorrect data into a domain type. If we ever have the need to do that it's probably an indicator that we did something wrong and should have handled an error in some infra module during deserialization.
But having this function `pub` seems a bit odd because that implies to me that it's intended to be a general purpose function that can be imported from somewhere else which doesn't seem to be the case here. Otherwise we probably run into the same issue of having to introduce a bunch of boiler plate code to convert the errors.